### ISSUE TITLE
v1.2.0: Prysm crashes on startup; illegal instruction

### ISSUE THREAD

**Issue by prestonvanloon** [prysmaticlabs/prysm#8410]
# 🐞 Bug Report

### Description

```
SIGILL: illegal instruction
PC=0x16f84f3 m=0 sigcode=2
signal arrived during cgo execution
instruction bytes: 0xf3 0x48 0xf 0x38 0xf6 0xc8 0x66 0x4d 0xf 0x38 0xf6 0xc6 0xc4 0x42 0xfb 0xf6

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x167aa80, 0xc00063dde8, 0x8)
        GOROOT/src/runtime/cgocall.go:133 +0x5b fp=0xc00063ddb8 sp=0xc00063dd80 pc=0x42417b
github.com/herumi/bls-eth-go-binary/bls._Cfunc_blsInit(0xf600000005, 0xc000000000)
        _cgo_gotypes.go:422 +0x4d fp=0xc00063dde8 sp=0xc00063ddb8 pc=0xcd5aed
github.com/herumi/bls-eth-go-binary/bls.Init(0x5, 0x0, 0x0)
        bls.go:63 +0x45 fp=0xc00063de40 sp=0xc00063dde8 pc=0xcdad65
github.com/prysmaticlabs/prysm/shared/bls/herumi.init.0()
        shared/bls/herumi/init.go:6 +0x2e fp=0xc00063de68 sp=0xc00063de40 pc=0xce47ee
runtime.doInit(0x29620e0)
        GOROOT/src/runtime/proc.go:5652 +0x8a fp=0xc00063de98 sp=0xc00063de68 pc=0x4681ca
runtime.doInit(0x295cc00)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063dec8 sp=0xc00063de98 pc=0x468197
runtime.doInit(0x296af00)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063def8 sp=0xc00063dec8 pc=0x468197
runtime.doInit(0x2970860)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df28 sp=0xc00063def8 pc=0x468197
runtime.doInit(0x29648c0)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df58 sp=0xc00063df28 pc=0x468197
runtime.doInit(0x296d8c0)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df88 sp=0xc00063df58 pc=0x468197
runtime.main()
        GOROOT/src/runtime/proc.go:191 +0x1c5 fp=0xc00063dfe0 sp=0xc00063df88 pc=0x45ae45
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc00063dfe8 sp=0xc00063dfe0 pc=0x491881
```

### Has this worked before in a previous version?

v1.1.0. Seems related to recent Herum updates.

## 🔬 Minimal Reproduction

TBD

## 🔥 Error

```
SIGILL: illegal instruction
PC=0x16f84f3 m=0 sigcode=2
signal arrived during cgo execution
instruction bytes: 0xf3 0x48 0xf 0x38 0xf6 0xc8 0x66 0x4d 0xf 0x38 0xf6 0xc6 0xc4 0x42 0xfb 0xf6

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x167aa80, 0xc00063dde8, 0x8)
        GOROOT/src/runtime/cgocall.go:133 +0x5b fp=0xc00063ddb8 sp=0xc00063dd80 pc=0x42417b
github.com/herumi/bls-eth-go-binary/bls._Cfunc_blsInit(0xf600000005, 0xc000000000)
        _cgo_gotypes.go:422 +0x4d fp=0xc00063dde8 sp=0xc00063ddb8 pc=0xcd5aed
github.com/herumi/bls-eth-go-binary/bls.Init(0x5, 0x0, 0x0)
        bls.go:63 +0x45 fp=0xc00063de40 sp=0xc00063dde8 pc=0xcdad65
github.com/prysmaticlabs/prysm/shared/bls/herumi.init.0()
        shared/bls/herumi/init.go:6 +0x2e fp=0xc00063de68 sp=0xc00063de40 pc=0xce47ee
runtime.doInit(0x29620e0)
        GOROOT/src/runtime/proc.go:5652 +0x8a fp=0xc00063de98 sp=0xc00063de68 pc=0x4681ca
runtime.doInit(0x295cc00)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063dec8 sp=0xc00063de98 pc=0x468197
runtime.doInit(0x296af00)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063def8 sp=0xc00063dec8 pc=0x468197
runtime.doInit(0x2970860)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df28 sp=0xc00063def8 pc=0x468197
runtime.doInit(0x29648c0)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df58 sp=0xc00063df28 pc=0x468197
runtime.doInit(0x296d8c0)
        GOROOT/src/runtime/proc.go:5647 +0x57 fp=0xc00063df88 sp=0xc00063df58 pc=0x468197
runtime.main()
        GOROOT/src/runtime/proc.go:191 +0x1c5 fp=0xc00063dfe0 sp=0xc00063df88 pc=0x45ae45
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc00063dfe8 sp=0xc00063dfe0 pc=0x491881
```

## 🌍  Your Environment

**Operating System:**

Users reporting on
- Linux hostname 4.19.0-11-amd64 #1 SMP Debian 4.19.146-1 (2020-09-17) x86_64 GNU/Linux
- Ubuntu 20.04.2 LTS (Focal Fossa)

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
v1.2.0
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

**Comment by prestonvanloon** [comment]
One users' CPU info.
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 63
model name      : Intel(R) Xeon(R) CPU E5-2637 v3 @ 3.50GHz
stepping        : 2
microcode       : 0x44
cpu MHz         : 1198.677
cache size      : 15360 KB
physical id     : 0
siblings        : 8
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 15
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg
 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_l
lc dtherm ida arat pln pts md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit
bogomips        : 6983.47
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
```

My CPU info (I can't reproduce on my machine):
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 142
model name      : Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
stepping        : 12
microcode       : 0xde
cpu MHz         : 1022.867
cache size      : 8192 KB
physical id     : 0
siblings        : 8
core id         : 0
cpu cores       : 4
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_g
ood nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand l
ahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap cl
flushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities
bugs            : spectre_v1 spectre_v2 spec_store_bypass swapgs itlb_multihit srbds
bogomips        : 3999.93
```

**Comment by prestonvanloon** [comment]
cc: @herumi

**Comment by yorickdowne** [comment]
Fine on `AMD EPYC 7282` and `E5-2630 v4`. 

Since an `E5-2637 v3` crashed, it's likely that the issue is with instructions missing in "Haswell" but present in "Broadwell"

These are them:

"
Broadwell introduces some instruction set architecture extensions:[16][17]

Intel ADX: ADOX and ADCX for improving performance of arbitrary-precision integer operations[18]
RDSEED for generating 16-, 32- or 64-bit random numbers from a thermal noise entropy stream, according to NIST SP 800-90B and 800-90C[19]
PREFETCHW instruction[19]
Supervisor Mode Access Prevention (SMAP) – optionally disallows access from kernel-space memory to user-space memory, a feature aimed at making it harder to exploit software bugs.[20][21]
Transactional Synchronization Extensions: This instruction set is reintroduced for all versions of Broadwell except for Broadwell-Y because a bug that cannot be fixed via microcode update in Broadwell-Y and all versions of Haswell except for the Haswell-EX variants has been fixed with a new CPU stepping level.[10] Erratum: In fact, among Broadwell i3, i5 and i7 CPUs, only four of them support TSX instructions (i7 5650U and 5600U, i5 5350U and 5300U); it is not even precised on Intel's website whether i5 5200U does support TSX instructions. (ark.intel.com/products/)
"

**Comment by yorickdowne** [comment]
Disassembly. It's ADOX and ADCX.

```
Raw Hex (zero bytes in bold):

F3480F38F6C8664D0F38F6C6C442FBF6   

String Literal:

"\xF3\x48\x0F\x38\xF6\xC8\x66\x4D\x0F\x38\xF6\xC6\xC4\x42\xFB\xF6"

Array Literal:

{ 0xF3, 0x48, 0x0F, 0x38, 0xF6, 0xC8, 0x66, 0x4D, 0x0F, 0x38, 0xF6, 0xC6, 0xC4, 0x42, 0xFB, 0xF6 }

Disassembly:
0:  f3 48 0f 38 f6 c8       adox   rcx,rax
6:  66 4d 0f 38 f6 c6       adcx   r8,r14
c:  c4                      .byte 0xc4
d:  42 fb                   rex.X sti
f:  f6                      .byte 0xf6
```

**Comment by prestonvanloon** [comment]
The issue is that the Herumi library was updated and in this new update, it uses CPU operations from Intel ADX. (https://en.wikipedia.org/wiki/Intel_ADX) This was introduced in Intel Broadwell (5th gen) chips (since discontinued in 2018). If your intel CPU is Haswell (4th gen) or older, you will not be able to run Prysm v1.2.0.

**Comment by prestonvanloon** [comment]
Looks like this did fail in Herumi's testing: https://travis-ci.org/github/herumi/bls-eth-go-binary/jobs/756558632

And then was disabled to pass: https://github.com/herumi/bls-eth-go-binary/commit/a603657044c41f231dafae00ee70d733cad667f0

**Comment by prestonvanloon** [comment]
How to determine your intel chip generation: https://www.intel.com/content/www/us/en/support/articles/000032203/processors/intel-core-processors.html

**Comment by herumi** [comment]
I fixed the bug at https://github.com/herumi/mcl/commit/b4abccd9a5581de28afbf3831b15e104b69c82b7 .
Could you try the latest [bls-eth-go-binary](https://github.com/herumi/bls-eth-go-binary/commit/8e0eb5237d01d8e53b57d45f6d12b0e86c299c7d)?

**Comment by prestonvanloon** [comment]
@herumi Thanks for quick fix. Working on updating and testing now.

**Comment by prestonvanloon** [comment]
I've uploaded binaries with commit 0b3ade3, tagged them as `v1.2.1-rc`, and uploaded them.

If anyone with Intel Haswell or older could run this release candidate, that would be really helpful.

```
USE_PRYSM_VERSION=v1.2.1-rc ./prysm.sh beacon-chain --pyrmont
```

**Comment by eth2353** [comment]
Hey Preston, just tried the RC, but it doesn't fix the issue on macOS/darwin, it still crashes immediately with SIGILL and the same instruction bytes as before.
I think it's because you used bls-eth-go-binary @ 8e0eb52 , and bls-eth-go-binary doesn't contain the binaries for android/mac/windows at that commit yet, see https://github.com/herumi/bls-eth-go-binary/commits/dev .

**Comment by herumi** [comment]
>it doesn't fix the issue on macOS/darwin
I've pushed the compiled binaries in order.

https://github.com/herumi/bls-eth-go-binary/releases/tag/v1.21 contains all compiled binaries.

**Comment by prestonvanloon** [comment]
Yes, I used  bls-eth-go-binary @ 8e0eb52 as that is what @herumi linked.

Thanks for confirming @eth2353. I will have to rebuild binaries and try again.

**Comment by prestonvanloon** [comment]
We've received some confirmation that this works for linux, but the previous candidate did not have darwin libraries updated.

I've uploaded another RC, `v1.2.1-rc.1`, built off commit 622901c. Please test this for darwin.

```
USE_PRYSM_VERSION=v1.2.1-rc.1 ./prysm.sh beacon-chain --pyrmont
```

**Comment by eth2353** [comment]
Confirming `v1.2.1-rc.1` now works for darwin.

**Comment by prestonvanloon** [comment]
This has been released in v1.2.1. 
Thanks for all of the help @herumi @eth2353 @yorickdowne
